internal/cluster/clusteradapter/distribution_eks_test.go

```diff
@@ -0,0 +1,149 @@
+// Copyright © 2020 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusteradapter
+
+import (
+	"context"
+	"testing"
+
+	"emperror.dev/errors"
+	"github.com/stretchr/testify/require"
+
+	"github.com/banzaicloud/pipeline/internal/cluster"
+	"github.com/banzaicloud/pipeline/internal/cluster/distribution/eks"
+)
+
+func TestEksServiceListNodePools(t *testing.T) {
+	type constructionArgumentType struct {
+		service eks.Service
+	}
+	type functionCallArgumentType struct {
+		ctx       context.Context
+		clusterID uint
+	}
+	type functionCallResultType struct {
+		expectedNodePools cluster.RawNodePoolList
+		isExpectingError  bool
+	}
+	type testCaseType struct {
+		caseName              string
+		constructionArguments constructionArgumentType
+		functionCallArguments functionCallArgumentType
+		functionCallResult    functionCallResultType
+		setupMockFunction     func(constructionArgumentType, functionCallArgumentType)
+	}
```

Comment:
While I understand what these structs are used for, I find it harder to read code like this.

For example: the testCaseType is never reused so it could be defined as an inline type in the test case list definition.

I guess this is a question of taste as well, but I usually prefer just plain local variables for the rest.

That being said, I'd be open to hearing more arguments while this style/practice is better from a structural point of view.

For now, it's okay.

Reply:
I have refined my approach in the [latest rebase](https://github.com/banzaicloud/pipeline/compare/e19e85be0dba082a8c755a58d53a4dfb4270fc43..f00f068d4e08f9e383f9d21c852421e0053ef3e2).

I kept the `constructionArgumentType` and `functionCallArgumentType` and eliminated/inlined the rest of the test types.
The reason behind these is the repeated use in the test case and the setup function so some syntactic code repetition can be omitted this way which IMO improves readability as only 2 parameters are passed into the `setupMockFunction` instead of arbitrary number of parameters - often 5+ - and also I like the origin/purpose encapsulation of this approach when reading the test code.

Let me know what you think about this mixed approach and whether I should push further and remove those two test type definitions as well.

Reply:
Looks good to me. I'm all for trying new approaches. If it indeed improves the readability/shortens the bootstrap code, let's do it!